enhancements/ingress/global-options-enable-hsts.md

```diff
@@ -0,0 +1,411 @@
+it never needs to do an initial load to get the HSTS header (note that this is not defined in RFC 6797
+and is therefore client implementation-dependent)
+
+HSTS is currently implemented as a pattern in the HAProxy template and applied to `edge` and
+`reencrypt` routes that match the `haproxy.router.openshift.io/hsts_header` annotation:
+```gotemplate
+{{- /* hsts header in response: */}}
+ {{- $hstsOptionalTokenPattern := `(?:includeSubDomains|preload)` }}
+ {{- $hstsPattern := printf `(?:%[1]s[;])*max-age=(?:\d+|"\d+")(?:[;]%[1]s)*`  $hstsOptionalTokenPattern -}}
+...
+{{- if matchValues (print $cfg.TLSTermination) "edge" "reencrypt" }}
+    {{- with $hsts := firstMatch $hstsPattern (index $cfg.Annotations "haproxy.router.openshift.io/hsts_header") }}
+  http-response set-header Strict-Transport-Security {{$hsts}}
+    {{- end }}{{/* hsts header */}}
+  {{- end }}{{/* is "edge" or "reencrypt" */}}
+```
+
+#### Enable HSTS for all routes globally, based on the cluster's ingress domain
+To add the header to all HTTP responses from an `edge` or `reencrypt` route, add a new router environment
+variable called `ROUTER_DEFAULT_HSTS`.
+
+Add a new type `IngressControllerHSTSHeaderPolicy` to the API in
+`openshift/api/blob/master/operator/v1/types_ingress.go`, to capture the configuration of the HSTS Policy.
+```go
+type IngressControllerHSTSHeaderPolicy {
+    // max-age is the delta time range in seconds during which the host is regarded as an HSTS host
+	// If set to 0, it negates the effect and the host is no longer regarded as an HSTS host.
+	// max-age` decrements over time.
+    // +required
+    MaxAge             string `json:"max-age"`
+
+    // scope indicates whether this header policy is enabled on a per-router basis (i.e. some
+    // routers will use HSTS and some will not) or it is global across every router in the cluster
+    // ingress domain.  If per-router HSTS configuration scope is desired, set the scope field to "limited".
+    // If global HSTS is desired, set the scope field to "all".  The default is "all".
+    // +optional
+    Scope               string `json:"scope"`
+
+    // inScopeRouters is a list of router domains that require HSTS, when not all routers require HSTS.
+    // If scope = all then this field is ignored.  If scope = limited, HSTS policy is limited to the routers
+    // matching the domains in the inScopeRouters list.
+    // +optional
+    InScopeRouters      []string 'json:"inScopeRouters'
+
+    // includeSubDomains means the HSTS Policy applies to any subdomains of the host's domain name
+    // +optional
+    IncludeSubDomains   bool `json:"includeSubDomains,omitempty"`
+
+    // preload tells the client to include this host in its host preload list so that
+    // it never needs to do an initial load to get the HSTS header (note that this is not defined
+    // in RFC 6797 and is therefore client implementation-dependent).
+    // +optional
+    Preload             bool `json:"preload,omitempty"`"`
```

Comment:
I'm tempted to suggest making this a list of attributes:
```go
    additionalAttributes []hstsAttribute `json:"additionalAttributes,omitempty"`"`
    // ...
}

// hstsAttribute is an attribute that can be specified on HTTP Strict-Transport-Security headers.
// +kubebuilder:validation:Enum=preload;includeSubDomains
type hstsAttribute string
```
What do you think?  Might want to check with someone with good taste and familiarity with API conventions.

Reply:
I felt wishy-washy that the three header components `max-age`, 'preload`, and 'includeSubDomains` did not belong in the same grouping as scope components `scope` and `inScopeDomains`, so I like the suggestion.  Maybe separating these into `headerDirectives` and `scopeAttributes`.  Would @deads2k be a good person to ask?

Reply:
Updated this to use a `hstsDirectives` list